### DISCUSSION TITLE
Can you please explain why in the below code async session is blocking

### DISCUSSION THREAD

**Discussion by cs-satish-mishra** [sqlalchemy/sqlalchemy#9871] [Usage Questions] [answered]
I have created a connection pool with max 1 connection can anyone please explain why my test1 function is blocking?  I was expecting that at asyncio.sleep connection should be released back to the pool and the test2 function should be able to continue. 
Please correct my understanding. 

```
import asyncio

from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker

engine = create_async_engine('postgresql+asyncpg://user:mypass@localhost:5432/test', pool_size=1, max_overflow=0,
                             echo=True)
                             
async_session = sessionmaker(
    engine, expire_on_commit=False, class_=AsyncSession
)


async def test1():
    async with async_session() as session:
        await session.execute('select 1')
        await asyncio.sleep(15)
        await session.execute('select 1')
        print('completed test1')


async def test2():
    async with async_session() as session:
        await session.execute('select 1')
        print('completes test2')


async def async_main():
    await asyncio.gather(test1(), test2())


if __name__ == "__main__":
    asyncio.run(async_main())
```

**Comment by cs-satish-mishra** [comment]
@zzzeek @CaselIT Please check when you can? Thanks

**Comment by CaselIT** [comment]
Seems to be working as expected. You place a sleep while keeping the session in test 1 active, so test2 need to wait for that to finish, since the pool has a single connection.

> I was expecting that at asyncio.sleep connection should be released back to the pool

Why would it do that? The session is what's controlling the connection, so until you exited the context manager noting it telling it to finish and release the connection

**Comment by cs-satish-mishra** [comment]
@CaselIT  Thanks for looking into it 
But if I change and put sleep first it does not block, does it mean connection is acquired at first execute and not as soon as we enter the context? 
```

async def test1():
    async with async_session() as session:

        print('starting test1')
        await asyncio.sleep(5)
        await session.execute('select 1')
        print('completed test1')
```

**Reply by CaselIT** [reply]
Yes, session acquires it lazily when it first needs it

**Comment by cs-satish-mishra** [comment]
@CaselIT The actual use case I am trying is to set the correct schema based on the 'schema' received in the API request's subdomain or host in a multitenant app based on PostgreSQL schema. For this, I am creating a pool of connections in Fastapi and looking for ways to set the schema dynamically when the request is made. I am trying to figure out if I could prepare a db query which is schema aware at runtime without blocking it in the session 

```
async def list(schema: str, session: AsyncSession):
    query = select(User)  # this query should be schema-aware eg if ischema is tenant1 then query should be tenant1.user.name from..
    return await session.execute(query) 
    ```

**Reply by CaselIT** [reply, marked as answer]
Not sure what sleep has to do with this or blocking, but you may use schema translate map https://docs.sqlalchemy.org/en/20/core/connections.html#translation-of-schema-names to change the schema of a model

**Comment by cs-satish-mishra** [comment]
This helps,  thanks
